### ISSUE TITLE
DatePicker: Accessibility 508 Compliance

### ISSUE THREAD

**Issue by kaushikmeduri** [primefaces/primefaces#4685]
## 1) Environment
- PrimeFaces version:6.1.19
- Does it work on the newest released PrimeFaces version? Version? No
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) No
- Application server + version:  Jboss 6.0
- Affected browsers: All browsers

## 2) Expected behavior

We are using https://www.primefaces.org/showcase/ui/input/datePicker.xhtml.
We are expecting it to be 508 / key board / screen reader compliant like
https://dequeuniversity.com/library/aria/date-pickers/sf-date-picker
...

## 3) Actual behavior

Date Picker is not accessibly via key board for date selections
..

## 4) Steps to reproduce

use any of https://www.primefaces.org/showcase/ui/input/datePicker.xhtml.
Date Picker is not accessibly via key board for date selections 
We are expecting it to be 508 / key board / screen reader compliant like
https://dequeuniversity.com/library/aria/date-pickers/sf-date-picker
## 5) Sample XHTML
<h:outputText  id="reportingPeriodStart"
	value="#{homeBean.startDate}" >
	<f:convertDateTime pattern="MM/dd/yyyy" />
</h:outputText>	
 

## 6) Sample bean

@ManagedBean(name = "homeBean")
@ViewScoped
public class HomeController {

private Date startDate;

public Date getStartDate() {
return startDate;
}

public void setStartDate(Date startDate) {
this.startDate = startDate;
}
}
..

**Comment by tandraschko** [comment]
Feel free to provide a PR

**Comment by kaushikmeduri** [comment]
We have Elite Account . Please advise what is PR and steps to create PR .

**Comment by tandraschko** [comment]
google about GitHub Pull Requests (PR)

**Comment by kaushikmeduri** [comment]
hi , We have not implemented any date pickers yet.
Do you know any prime faces extension some one already have for this ..

**Comment by kaushikmeduri** [comment]
We do not have fix . Please advise if prime faces can provide a price quote for fixes to be included as patch. In the past our agency has some work done for 508 fixes

**Comment by tandraschko** [comment]
Please see: primefaces.org/support - there is a "get a quote" button

**Comment by christophs78** [comment]
As a reference what others do: https://material.angular.io/components/datepicker/overview#keyboard-interaction

**Comment by christophs78** [comment]
We may try to backport primefaces/primeng#4695, primefaces/primeng@ea9596b from PrimeNG for DatePicker. (DatePicker is the replacement for Calendar)
It´s a bit sad PrimeTek only added this for PrimeNG, PrimeReact and PrimeVue. Not much love for PrimeFaces.

**Comment by melloware** [comment]
Agreed.  And keyboard accessibility would be a huge feature.  I feel like we need to start monitoring the PrimeNG issues for features that get added so we can try and keep PF in sync.

**Comment by abhishar03** [comment]
Is this Still open , same issues

**Comment by melloware** [comment]
@abhishar03 Yes this ticket is still open and has not been implemented yet.

**Comment by efibn** [comment]
This issue really have to get prioritized, PLEASE !

**Comment by melloware** [comment]
@efibn your best bet might be to pay for a 1 time PRO support to PrimeTek and fund this issue.

**Comment by efibn** [comment]
@yigitfindikli Please take a look at this, if you open calendar from dialog it cannot be controlled by the keyboard, for accessibility purposes, same goes for Calendar that is in between dates.
Please take a look at this

**Comment by melloware** [comment]
@efibn @yigitfindikli is the PrimeNG lead and this is the PrimeFaces ticket.  Do you have the right issue tracker?

**Comment by Bukama** [comment]
The problem that the DatePicker is not accessible via keyboard, was also a downrating in the accessibilty tests of our web-apps. As it is planed that Primefaces 13 should focus on accessibility (according to Cagatay Civici) I would like to ask the whole team at PrimeTek to give this issue the needed ressourced to get fixed.

**Comment by jepsar** [comment]
> The problem that the DatePicker is not accessible via keyboard, was also a downrating in the accessibilty tests of our web-apps. As it is planed that Primefaces 13 should focus on accessibility (according to Cagatay Civici) I would like to ask the whole team at PrimeTek to give this issue the needed ressourced to get fixed.

Agreed, but keep in mind that this is the community edition issue tracker. https://github.com/primefaces/primefaces#community--elite--pro

**Comment by Bukama** [comment]
> Agreed, but keep in mind that this is the community edition issue tracker. 

I'm totally aware of this. Nevertheless I hope that (and warmly ask) the PrimeTek team has the same opinion that the DatePicker is a largely used component which should be accessible and decides therefore to put "more than regular" ressource in the upcoming version, when accessibility is promoted widely for all PrimeTek libaries. Maybe (if technically possible) reusing code or even ressources from PrimeNG lib, which copyrights are also at PrimeTek might reduce the overall costs for reaching the goal. 

So please don't see my comment as a "I WANT YOU TO DO THIS! NOW!"-claim ❤

**Comment by melloware** [comment]
as far as I know the PrimeVue Calendar is now 100% accessible: https://www.primefaces.org/primevue/calendar

Those fixes will eventually be ported to our PF DatePicker.

**Comment by ZeroBrushV2** [comment]
> as far as I know the PrimeVue Calendar is now 100% accessible: https://www.primefaces.org/primevue/calendar
> 
> Those fixes will eventually be ported to our PF DatePicker.

I would like to mention also the PrimeReact Calendar is also 100% accessible: https://www.primefaces.org/primereact/calendar/

**Comment by melloware** [comment]
I started working on this and have been making it match all the accessibility changes from PrimeVue.  Making good progress so far.

**Comment by tarwork** [comment]
@melloware  thank you for working on this. We are facing this exact issue literally right now in getting an app past our internal accessibility audit.

We have the ELITE license so hopefully we can get a point release with the fix ?

**Comment by melloware** [comment]
@tarwork after the team thoroughly tests my changes (next couple of weeks) I can ask PrimeTek about porting to to an Elite release.

**Comment by Bukama** [comment]
> @tarwork after the team thoroughly tests my changes (next couple of weeks) I can ask PrimeTek about porting to to an Elite release.

It was announced that ELITE and Community are combined to a free LTS since 13.0.0, so should be available for all, right?

**Comment by melloware** [comment]
@Bukama you are totally right but I just worry because so far I have been patching 13.0.1 with bugfixes only this is a new feature/breaking change.   But we will get more clarity from PrimeTek and the other team members before we do anything.

**Comment by fkalanti** [comment]
Hi,
this is a really often needed feature, so if i understand correctly only available from Primefaces Version 14 on? 
Does the possibility exist to make the DatePicker accessible also in Version 13 of Primefaces?
Thank you!

**Comment by melloware** [comment]
Unfortunaltey @fkalanti this was such a massive change to the component there is no way it was just a "backpatch" to 13.  So 14.0.2 is your best best if you want this feature.
